mixer/pkg/runtime2/routing/builder.go

```diff
@@ -0,0 +1,376 @@
+// Copyright 2018 Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package routing implements a routing table for resolving incoming requests to handlers. The table data model
+// is structured for efficient use by the runtime code during actual dispatch. At a high-level, the structure
+// of table is as follows:
+//
+// Table:          map[variety]varietyTable
+// varietyTable:   map[namespace]NamespaceTable
+// NamespaceTable: list(Destination)
+// Destination:    unique(handler&template) + list(InputGroup)
+// InputGroup:     condition + list(InstanceBuilders) + list(OutputMappers)
+//
+// The call into table.GetDestinations performs a lookup on the first map by the variety (i.e. quota, check,
+// report, apa etc.), followed by a lookup on the second map for the namespace, and a NamespaceTable struct
+// is returned. (If a namespace is not resolved, then data for the default namespace is used).
+//
+// The NamespaceTable holds all the handlers that should be dispatched to, within this namespace, along with
+// conditions and builders for the instances. The same handler can appear multiple times in this list, for
+// each template that is supported by the handler. This helps caller to ensure that each dispatch to the
+// handler will use a unique template.
+//
+// The client code is expected to work as follows:
+// - Call GetDestinations(variety, namespace) to get a NamespaceTable.
+// - Go through the list of entries in the NamespaceTable.
+// - For each entry begin a dispatch session to the associated handler.
+// - Go through the InputGroups
+// - For each InputGroup, check the condition and see if the inputs/outputs apply.
+// - If applies, then call InstanceBuilders to create instances
+// - Depending on the variety, either aggregate all instances in the group, and send them all at once, or
+//   dispatch for every instance individually to the adapter.
+//
+package routing
+
+import (
+	"fmt"
+	"strings"
+
+	descriptor "istio.io/api/mixer/v1/config/descriptor"
+	tpb "istio.io/api/mixer/v1/template"
+	"istio.io/istio/mixer/pkg/adapter"
+	"istio.io/istio/mixer/pkg/expr"
+	"istio.io/istio/mixer/pkg/il/compiled"
+	"istio.io/istio/mixer/pkg/runtime2/config"
+	"istio.io/istio/mixer/pkg/runtime2/handler"
+	"istio.io/istio/mixer/pkg/template"
+	"istio.io/istio/pkg/log"
+)
+
+// builder keeps the ephemeral state while the routing table is built.
+type builder struct {
```

Comment:
can you separate datapath components from configuration path components?
I though runtime2.routing is a datapath package, but it is a mix.

Reply:
I don't think that is good dimension to factor on. runtime2.routing contains a data structure, along with its builder. The builder can create things in the data structure that are not (or should not) necessarily be visible to the client code.